### ISSUE TITLE
Rule request: Require uppercase variables to be typed with `typing.Final`

### ISSUE THREAD

**Issue by adamtheturtle** [astral-sh/ruff#10137]
In my projects, when I spot an uppercase variable, I make sure to type it with `typing.Final`.

It would be nice to have this as a ruff rule.

For example, from the [Python documentation's example of `typing.Final`](https://docs.python.org/3/library/typing.html#typing.Final):

```python
MAX_SIZE: Final = 9000
MAX_SIZE += 1  # Error reported by type checker
```

I'd like:

```python
MAX_SIZE = 9000 # ruff shows an error because this is not typed with `typing.Final`.
```

**Comment by AlexWaygood** [comment]
I actually suggested something like this a while back for flake8-pyi: https://github.com/PyCQA/flake8-pyi/issues/122. But we rejected the idea because it would likely cause too many false positives for stub files, for which the most important thing is the need to exactly follow what the runtime does.

For runtime Python code, there would likely be a lower number of false positives than in `.pyi` files. I think it would be a useful lint.

Not sure where it would belong -- it does feel like it fits in pretty well with flake8-pyi, but it would be the first rule in that ruleset that would run on `.py` files and not `.pyi` files 😄

**Comment by MichaReiser** [comment]
This was also discussed as a ty lint rule https://github.com/astral-sh/ty/issues/3650
